Pcap++/src/PcapFileDevice.cpp

```
@@ -131,6 +336,45 @@ namespace pcpp
 		return new PcapFileReaderDevice(fileName);
 	}
 
+	std::unique_ptr<IFileReaderDevice> IFileReaderDevice::createReader(const std::string& fileName)
+	{
+		std::ifstream fileContent(fileName, std::ios_base::binary);
+		if (fileContent.fail())
+		{
+			throw std::runtime_error("Could not open: " + fileName);
+		}
+
+		switch (CaptureFileFormatDetector().detectFormat(fileContent))
+		{
+		case CaptureFileFormat::PcapNano:
+		{
+			if (!checkNanoSupport())
+			{
+				PCPP_LOG_ERROR("Pcap nano second precision files are not supported in this build of PcapPlusPlus");
+				return nullptr;
+			}
+			// fallthrough
+		}
+		case CaptureFileFormat::Pcap:
+			return std::make_unique<PcapFileReaderDevice>(fileName);
+		case CaptureFileFormat::PcapNGZstd:
+		{
+			if (!checkZstdSupport())
+			{
+				PCPP_LOG_ERROR("PcapNG Zstd compressed files are not supported in this build of PcapPlusPlus");
+				return nullptr;
+			}
+			// fallthrough
+		}
+		case CaptureFileFormat::PcapNG:
+			return std::make_unique<PcapNgFileReaderDevice>(fileName);
+		case CaptureFileFormat::Snoop:
+			return std::make_unique<SnoopFileReaderDevice>(fileName);
+		default:
+			return nullptr;
+		}
+	}
```

Curious if it isn't simpler to throw on unsupported format, instead of returning `nullptr`... It would give the caller only 1 error mechanism to worry about.

Generally those are unrecoverable errors for the context, and we already throw on missing file?
If needed we can provide a second `tryCreate` factory for non-exceptional error flow.

---

Updated 45ad7695f0b73de6ccc3669666d48c6985523c1f
`tryCreateReader` has a dirty implementation, but it is not supposed to be used in performance heavy code. How often do people create new devices?

---

I wouldn't throw an exception for unsupported format. Unlike other languages, using `try..catch` in C++ isn't very common, mostly because it's a costly operation. Returning `nullptr` is better in this case